pkg/agent/nodeportlocal/nplagent_test.go

```diff
@@ -94,47 +119,244 @@ func TestMain(t *testing.T) {
 	mockCtrl := gomock.NewController(t)
 	portTable = NewPortTable(mockCtrl)
 
-	c, _ := InitController(kubeClient, portTable, defaultNodeName)
+	// informerFactory is initialised and started from cmd/antrea-agent/agent.go
+	informerFactory := informers.NewSharedInformerFactory(kubeClient, resyncPeriod)
+
+	c, _ := InitController(kubeClient, informerFactory, portTable, defaultNodeName)
 	stopCh := signals.RegisterSignalHandlers()
-	go c.Worker()
-	go c.Ctrl.Run(stopCh)
-	cache.WaitForCacheSync(stopCh, c.Ctrl.HasSynced)
+
+	go c.Run(stopCh)
+	informerFactory.Start(stopCh)
+}
+
+func pollForPodAnnotation(r *require.Assertions, podName string, found bool) (string, error) {
+	var data string
+	var exists bool
+	err := wait.Poll(time.Second, 20*time.Second, func() (bool, error) {
+		updatedPod, err := kubeClient.CoreV1().Pods(defaultNS).Get(context.TODO(), podName, metav1.GetOptions{})
+		r.Nil(err, "Failed to get Pod")
+		ann := updatedPod.GetAnnotations()
+		data, exists = ann[k8s.NPLAnnotationKey]
+		if found {
+			return exists, nil
+		} else {
+			return !exists, nil
+		}
+	})
+
+	return data, err
+}
+
+// Add a Service (proper annotation) and Pod pair, update Service with bad
+// annotation value and check for Pod annotation and port cache entry removal.
+func TestSvcUpdateAnnotation(t *testing.T) {
+	var data string
+
+	a := assert.New(t)
+	r := require.New(t)
+
+	servicePodName := "pod-svc"
+
+	testSvc := getTestSvc()
+	testSvc.Name = servicePodName
+	testSvc.Spec.Selector[defaultAppSelectorKey] = servicePodName
+	testSvc.Annotations = map[string]string{k8s.NPLEnabledAnnotationKey: "true"}
+	s, err := kubeClient.CoreV1().Services(defaultNS).Create(context.TODO(), &testSvc, metav1.CreateOptions{})
+	r.Nil(err, "Service creation failed")
+	t.Logf("successfully created Service: %v", s)
+
+	testPod := getTestPod()
+	testPod.Name = servicePodName
+	testPod.Labels[defaultAppSelectorKey] = servicePodName
+	p, err := kubeClient.CoreV1().Pods(defaultNS).Create(context.TODO(), &testPod, metav1.CreateOptions{})
+	r.Nil(err, "Pod creation failed")
+	t.Logf("successfully created Pod: %v", p)
+
+	data, err = pollForPodAnnotation(r, servicePodName, true)
+	r.Nil(err, "Poll for annotation check failed")
+
+	var nplData []k8s.NPLAnnotation
+	json.Unmarshal([]byte(data), &nplData)
+
+	a.Len(nplData, 1)
+	a.Equal(nplData[0].NodeIP, defaultHostIP)
+	a.Equal(portTable.RuleExists(defaultPodIP, defaultPort), true)
+
+	// Update Service with bad annotation value.
+	testSvc.Annotations = map[string]string{k8s.NPLEnabledAnnotationKey: "false"}
+	testSvc.ResourceVersion = "2"
```

Comment:
is explicitly specifying the `ResourceVersion` in tests still needed with the latest version of the controller code?

Reply:
correct, Done